### ISSUE TITLE
TODO list for Transformer.

### ISSUE THREAD

**Issue by lcy-seso** [PaddlePaddle/Paddle#7355]
- [x] implement they layer normalization operator and the Python wrapper.
    - [x] CPU implementation.
    - [x] GPU implementation.
    - [x] python wrapper.
- [x] enhance the matmul operator to support 4-D tensor as its inputs https://github.com/PaddlePaddle/Paddle/issues/7319. fixed by PR: https://github.com/PaddlePaddle/Paddle/pull/7656
- [x] prepare [the dataset](https://arxiv.org/abs/1605.00459).  
 fixed by PR: https://github.com/PaddlePaddle/Paddle/pull/7661
- [x] wrap **the masked** positional embedding.
- [x] enhance the lookup_table operator to support the special token: padding index.  https://github.com/PaddlePaddle/Paddle/issues/7309. 
- [x] wrap the multi-head dot product attention. This is different to ConvS2S.
- [x] wrap the positional-wise feed-forward network. 
- [x] wrap the basic computation block.
- [x] build the entire model.
- [ ] enhance the documentation of operators used in Transformer.
- [x] add beam search for Transformer.
- [x] clean codes and merge the entire project into the [models](https://github.com/PaddlePaddle/models) repo (actually can merge the work part by part).
- [x] Learning Rate Scheduler
- [x] Residual Dropout
- [x] Label Smoothing
    - [x] label smooth operator.
    - [x] python wrapper.
- [x] Scaled Dot Product Attention
- [x] Weight sharing between embedding and pre-softmax linear transformation layers

**Comment by lcy-seso** [comment]
Hi, @abhinavarora and @kavyasrinet  I just add this todo list according to my understanding. Please feel free to make any modification to this list, and arrange the work as you needed.

Some steps are shared between ConvS2S and Transformer. The work has not been assigned yet.

**Comment by kavyasrinet** [comment]
Thank you so much @lcy-seso for compiling the list above. We were planning to work on this compilation on our end as well. We will update this soon, and add the additional items to the list above.
